### PR TITLE
adding required methods and functions for "__typename must be const" change  

### LINKED ISSUE #122: __typename must be const
Currently we allow this, but it's not technically safe since the `__typename` could be mutated.
```ts
class User implements IPerson {
  __typename = "User";
}
```

We should require either:

```ts
class User implements IPerson {
  __typename = "User" as const;
}
```

or 

```ts
class User implements IPerson {
  __typename: "User" = "User";
}
```

### PR DESCRIPTION
**Pull request by Rash-Hit**
Resolves #122 

This pr will add 
 - [x] Required Method and function for  " __typename must be const" Change
 - [x] Test Files 
 - [x] Changes In Other Test Files 

 
Before - 

image

After 

image

Fix 

image

--- 
@captbaritone  Please review , and let me know for the change's

### CODE REVIEW

**Comment by captbaritone** [comment]
Nice work, this is looking great! Next steps would be:

1. Update the existing tests (unit and itegration) to not trigger this issue
2. Create new test fixtures which trigger each of these conditions
3. Update docs to always include `as const` (Running `pnpm run build` in the grats dir and `pnpm run grats` from within the website dir should uncover some of these, as it will rebuild many of the snippets) 
4. Update example projects to include `as const` (`pnpm run integration-tests` should uncover these)

Tips on running fixtures tests can be found here: https://github.com/captbaritone/grats/blob/main/CONTRIBUTING.md#automated-tests

**Review by captbaritone — CHANGES_REQUESTED** [review]
Looking good! A few comments about error messages.

Also, this would be a great candidate for a code action. I landed support for them yesterday: https://github.com/captbaritone/grats/commit/0b53e3df0c583e650dd409f182d1f82c1df1926c

If you'd like to add support for that as part of this change, you can see an example [here](https://github.com/captbaritone/grats/blob/bdbc112016461f33a547c50a2aad06c5a5fe6d31/src/Extractor.ts#L219). (you'll need to rebase)

Also happy to add this myself or have that be a followup PR.

Finally, can search through the `website` directory for instances of `__typename` and see if there are any changes that need to be made there?

**Comment by captbaritone** [comment]
Also, can we make sure to have a fixture test for every error case? I don't think I see some of the new error messages represented in the test fixtures.

File: src/Extractor.ts [hunk 1]
```diff
@@ -740,18 +740,39 @@ class Extractor {
...
+      this.report(node.initializer.type, E.typeNameTypeNotReferenceNode());
       return false;
     }
+
+    if (!ts.isIdentifier(node.initializer.type.typeName)) {
+      this.report(node.initializer.type.typeName, E.typeNameTypeNameNotIdentifier());
+      return false;
+    }
+
+    if (node.initializer.type.typeName.escapedText !== "const") {
```
[the remark this answers is not in the document]
Comment (Rash-Hit) [reply src/Extractor.ts:881]: I've tried this thing , but getting this error 
image

and the reason is 
image

image

[the remark this answers is not in the document]
Comment (captbaritone) [reply src/Extractor.ts:881]: Interesting. I guess it doesn't parse `as const` as a keyword! Leaving this as a string is fine.

**Comment by Rash-Hit** [comment]
@captbaritone . sir can you please review it and let me know for the required changes ..

**Comment by captbaritone** [comment]
Thanks! I had some commits on main which I had to revert (unrelated to this PR) which meant I had to manually merge as https://github.com/captbaritone/grats/commit/1eed52344bd8061ceef7783170d9affcf42aea7c

You can see your commits (plus a few additions/and fixups) there.

**Pull request closed without merging**